File: 본문/python/Code_4_05_3.py
import queue

# 입력
N, M = map(int, input().split())
A = [ None ] * M
B = [ None ] * M
for i in range(M):
	A[i], B[i] = map(int, input().split())

# 인접리스트 작성
G = [ list() for i in range(N + 1) ]
for i in range(M):
	G[A[i]].append(B[i])
	G[B[i]].append(A[i])

# 너비 우선 탐색 초기화(dist[i] = -1일 때 아직 도달하지 않은 흰색 정점으로 취급)
dist = [ -1 ] * (N + 1)
Q = queue.Queue()
dist[1] = 0
Q.put(1) # Q에 1을 추가(조작1)

# 너비 우선 탐색
while not Q.empty():
	pos = Q.get() # Q의 앞을 확인하고, 이를 꺼냄(조작2, 3)
	for nex in G[pos]:
		if dist[nex] == -1:
			dist[nex] = dist[pos] + 1
			Q.put(nex) # Q에 nex를 추가(조작1)

# 정점1에서 각 정점까지의 최단 거리를 출력
for i in range(1, N + 1):
	print(dist[i])

File: 본문/cpp/Code_4_05_3.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

int N, M, A[100009], B[100009];
int dist[100009];
vector<int> G[100009];

int main() {
	// 입력
	cin >> N >> M;
	for (int i = 1; i <= M; i++) {
		cin >> A[i] >> B[i];
		G[A[i]].push_back(B[i]);
		G[B[i]].push_back(A[i]);
	}

	// 너비 우선 탐색 초기화(dist[i] = -1일 때, 미도달 흰색 정점으로 판단)
	for (int i = 1; i <= N; i++) dist[i] = -1;
	queue<int> Q; // 큐 Q 정의하기
	Q.push(1); dist[1] = 0; // Q에 1을 추가(조작1)

	// 너비 우선 탐색
	while (!Q.empty()) {
		int pos = Q.front(); // Q의 앞부분 확인(조작2)
		Q.pop(); // Q의 앞부분 추출(조작3)
		for (int i = 0; i < (int)G[pos].size(); i++) {
			int nex = G[pos][i];
			if (dist[nex] == -1) {
				dist[nex] = dist[pos] + 1;
				Q.push(nex); // Q에 nex를 추가(조작1)
			}
		}
	}

	// 정점1에서 각 정점까지의 최단 거리를 출력
	for (int i = 1; i <= N; i++) cout << dist[i] << endl;
	return 0;
}